mrmustard/lab_dev/sampler.py

```diff
@@ -0,0 +1,190 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Samplers for measurement devices.
+"""
+
+from __future__ import annotations
+from itertools import product
+
+from abc import ABC, abstractmethod
+
+from typing import Any, Sequence
+
+import numpy as np
+
+from mrmustard import math, settings
+
+from .states import State, Number, QuadratureEigenstate
+from .circuit_components import CircuitComponent
+from .circuit_components_utils import BtoQ
+
+__all__ = ["Sampler", "PNRSampler", "HomodyneSampler"]
+
+
+class Sampler(ABC):
+    r"""
+    A sampler for measurements of quantum circuits.
+
+    Args:
+        meas_outcomes: The measurement outcomes for this sampler.
+        povms: The POVMs of this sampler.
```

Comment:
`meas_outcomes` are just labels for the `povms` right? But since `povms` are instances of `CircuitComponent`, they already each have a `name` attribute. In the future, should we just rely on that attribute to label the outcomes?

Reply:
I would prefer not to rely on name particularly because relying on strings can be quite error prone. I also make a lot of use of this structure in the follow up measurement devices PR so it actually ends up being quite useful.